### ISSUE TITLE
Feature: inline classes

### ISSUE THREAD

**Issue by alisabzevari** [mockk/mockk#152]
I am trying to use Kotlin's experimental `inline class` feature. I have an interface with a method accepting a parameter with an inline class type. In this case, `every` function throws an exception.

### Prerequisites
My Kotlin version is: 1.3.0-rc-146

- [x] I am running the latest version - 1.8.9.kotlin13
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Expected Behavior

I should be able to define `every` with `any()` in parameters.

### Current Behavior

Test throws an exception.

### Failure Information (for bugs)

See the exception details in Failure logs section.

#### Steps to Reproduce

This is a test case. I am running this with `Kotlin 1.3.0-rc-146`, `kotlin-reflect 1.3.0-rc-146`,  `JUnit 5.1.0`, `mockk 1.8.9.kotlin13`.

``` kotlin
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Test

class MockkFailingTest {
    @Test // fails
    fun `mockk test using any()`() {
        val mocked = mockk<TestMock>(relaxed = true)
        every { mocked.test(any()) } returns 1
    }

    @Test // passes
    fun `mockk test using hardcoded value`() {
        val mocked = mockk<TestMock>(relaxed = true)
        every { mocked.test(MyInlineType("123")) } returns 1
    }

}

inline class MyInlineType(val value: String)

interface TestMock {
    fun test(value: MyInlineType): Int
}
```

#### Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

* MockK version: 1.8.9.kotlin13
* OS: Ubuntu 18.04
* Kotlin version: 1.3.0-rc-146
* JDK version: 1.8
* JUnit version: 5.1.0
* Type of test: unit test

#### Failure Logs

```
io.mockk.MockKException: Failed matching mocking signature for
SignedCall(retValue=0, isRetValueMock=false, retType=class kotlin.Int, self=TestMock(#2), method=test-gNmuAAA(String), args=[null], invocationStr=TestMock(#2).test-gNmuAAA(null))
left matchers: [any()]

	at io.mockk.impl.recording.SignatureMatcherDetector.detect(SignatureMatcherDetector.kt:99)
	at io.mockk.impl.recording.states.RecordingState.signMatchers(RecordingState.kt:38)
	at io.mockk.impl.recording.states.RecordingState.round(RecordingState.kt:30)
	at io.mockk.impl.recording.CommonCallRecorder.round(CommonCallRecorder.kt:45)
	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:47)
	at io.mockk.impl.eval.EveryBlockEvaluator.every(EveryBlockEvaluator.kt:25)
	at io.mockk.MockKDsl.internalEvery(API.kt:93)
	at io.mockk.MockKKt.every(MockK.kt:104)
	at conduit.handler.LoginHandlerImplTest.mockk test(LoginHandlerImplTest.kt:79)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:436)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:170)
	at org.junit.jupiter.engine.execution.ThrowableCollector.execute(ThrowableCollector.java:40)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:166)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:113)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:58)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:112)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:55)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:43)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:74)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

**Comment by oleksiyp** [comment]
Thanks for reporting it.

**Comment by manusoft** [comment]
`verify` throws same exception

**Comment by manusoft** [comment]
And cannot capture an inline class using `slot` either.

**Comment by RiccardoM** [comment]
To everyone that might come back to this issue, I found the way to make this work. 

Before writing down the solution, I want to make clear the following points: 
* mock checks are performed using the `.equals()` method, which is the same thing as `==` in Kotlin
* inline classes are only wrappers around a specific other class (in this case, `MyInlineType` is a wrapper around `String`) 
* inline classes' `.equals()` method just calls the wrapped value's `.equals()` method

With this in mind, we can re-write the following tests: 

```kotlin
@Test // fails
fun `mockk test using any()`() {
    val mocked = mockk<TestMock>(relaxed = true)
    every { mocked.test(any()) } returns 1
}
```

As the following

```kotlin
@Test // fails
fun `mockk test using any()`() {
    val mocked = mockk<TestMock>(relaxed = true)
    every { mocked.test(MyInlineType(any())) } returns 1
}
```

This will work, as the `MyInlineType` will be un-wrapped and the internal value will be checked without a problem later on. 

I hope this is useful to everyone. 

@oleksiyp Probably this might be wrote down into the `README` after more tests have been made.

**Comment by mkobit** [comment]
@RiccardoM that looks like a good workaround for the matching use case. What about the _"method returns an `inline class`"_ use case?

For example:

```kotlin
inline class TestMock(val value: String)

interface MyInterface {
  fun returnsMock(): TestMock
}
```

And tests:

```kotlin
internal class MockkInlineTest {

  @Test
  internal fun `returns example`() {
    val myInterface = mockk<MyInterface>()

    every { myInterface.returnsMock() } returns TestMock("hello")

    assertEquals(
      TestMock("hello"),
      myInterface.returnsMock()
    )
  }

  @Test
  internal fun `answers lambda example`() {
    val myInterface = mockk<MyInterface>()

    every { myInterface.returnsMock() } answers { TestMock("hello") }

    assertEquals(
      TestMock("hello"),
      myInterface.returnsMock()
    )
  }

  @Test
  internal fun `answers implementation example`() {
    val myInterface = mockk<MyInterface>()

    val answer = object : Answer<TestMock> {
      override fun answer(call: Call): TestMock {
        return TestMock("hello")
      }
    }
    every { myInterface.returnsMock() } answers(answer)

    assertEquals(
      TestMock("hello"),
      myInterface.returnsMock()
    )
  }
}
```

On Java 11 and Kotlin 1.3.31, all of these fail with a similar variant to:

```
java.lang.ClassCastException: class TestMock cannot be cast to class java.lang.String (TestMock is in unnamed module of loader 'app'; java.lang.String is in module java.base of loader 'bootstrap')

	at MyInterface$Subclass0.returnsMock(Unknown Source)
	at MockkInlineTest.answers implementation example
```

In Mockito, `answer` seemed to work as a workaround (that I found in https://github.com/nhaarman/mockito-kotlin/issues/309), but it doesn't seem to be working here. Any suggestions would be helpful.

**Comment by mkobit** [comment]
I don't believe this should be closed (either is or other issue) - can somebody add the `important` label?

**Comment by krzema12** [comment]
I use [klock](https://github.com/korlibs/klock) library which has an inline class at its core by design. I wanted to use mockk, but the lack of this feature/fix is a blocker to me to write unit tests.

If mockk doesn't want to miss new clients and lose existing ones, I suggest providing a complete workaround or even ugly, experimental API for this, and introduce proper API/fix in implementation later. Otherwise I and probably a bunch of other people will have to look for another library.

I understand that inline classes are an experimental feature of Kotlin and clients of libraries can expect undefined behavior in such case, but looks like authors of some libraries decide to rely on the experimental features really deeply and then it sort of forces other libraries to treat it in a "less experimental" way.

**Comment by ligi** [comment]
Thanks @krzema12 - your workaround worked for me and saved my day

unfortunately it made the code quite ugly - so I would really like to see this fixed.

**Comment by Pitel** [comment]
I have the same problem, but all the workarounds here are focusing on returning inline classes. But I need to verify the call of method with inline class as it's parameter.

```kotlin
inline class Group(val uuid: UUID) {
	constructor(uuid: String) : this(UUID.fromString(uuid))
}
```

And my test looks like this:

```kotlin
private companion object {
	private val GROUP_UUID = Group("4566900c-91b2-43cc-953e-28fe6354bf57")
}

@Test
fun testSomething() = runBlocking {
	doSomething()
	coVerify { foo.bar(GROUP_UUID) }
}
```

And this fails with:

```
SignedCall(..., method=bar-x2tluNw(UUID, Continuation), args=[null, Continuation at ...], invocationStr=Foo(foo#61).bar-x2tluNw(null, continuation {}))
```

How come that the UUID/Group parameter is `null`?!

**Comment by underlow** [comment]
Proposed workaround doesn't work with `kotlin.time.Duration` because Duration ctor is internal and 
```
 every { mocked.test(Duration(any())) } returns 1
```
impossible.

I've tried 

```
every { mocked.test(any<Double>().seconds) } returns 1
```
but no luck.

Any ideas?

**Comment by Zhelyazko** [comment]
I'm also having troubles with `kotlin.time.Duration` because of its internal constructor. Any fixes or workarounds?

**Comment by blazeroni** [comment]
I found a workaround for inline classes with inaccessible constructors like `kotlin.time.Duration`: make the constructor accessible 😮

Define a function like this:
```
fun MockKMatcherScope.anyDuration(): Duration {
    val ctor = Duration::class.constructors.first()
    ctor.isAccessible = true

    return ctor.call(any<Double>())
}
```

Usage:
```
every { mocked.test(anyDuration()) } returns 1
```

**Comment by qoomon** [comment]
@blazeroni A slightly more generic Workaround
```Kotlin
inline fun <reified T : Any, reified I : Any> MockKMatcherScope.anyInline() =
    T::class.constructors.first()
        .apply { isAccessible = true }
        .call(any<I>())
```

**Comment by qoomon** [comment]
@blazeroni a even more generic version.
```kotlin
inline fun <reified T : Any> MockKMatcherScope.anyInline(): T {
    return T::class.primaryConstructor!!.run {
        val valueType = parameters[0].type.classifier as KClass<*>
        val callRecorder = getProperty("callRecorder") as MockKGateway.CallRecorder
        val any = callRecorder.matcher(ConstantMatcher<T>(true), valueType)
        call(any)
    }
}
```

**Comment by qoomon** [comment]
@blazeroni and finally a total replacement
```kotlin
inline fun <reified T : Any> MockKMatcherScope.anyValue(): T =
    if (T::class.isInline) {
        T::class.primaryConstructor!!.run {
            call(anyValue(parameters[0].type.classifier as KClass<*>))
        }
    } else any()

fun <T : Any> MockKMatcherScope.anyValue(cls: KClass<T>): T =
    (getProperty("callRecorder") as MockKGateway.CallRecorder)
        .matcher(ConstantMatcher<T>(true), cls)

val KClass<*>.isInline: Boolean
    get() = !isData &&
        primaryConstructor?.parameters?.size == 1 &&
        java.declaredMethods.any { it.name == "box-impl" }
```

**Comment by qoomon** [comment]
however inline class as return values still not working :-/
`... returns 3.days` gives following exception at runtime 
`java.lang.ClassCastException: class kotlin.time.Duration cannot be cast to class java.lang.Double (kotlin.time.Duration is in unnamed module of loader 'app'; java.lang.Double is in module java.base of loader 'bootstrap')`

**Comment by qoomon** [comment]
finally I got it working even for return values see code below

**Usage Example for inline class kotlin.time.Duration**
```kotlin
mockk<Dummy> {
    every { functionWithDurationParameter(anyValue()) } returns value(3.days)
}
```

**Mockk Extension Functions**
```kotlin
import io.mockk.ConstantMatcher
import io.mockk.Matcher
import io.mockk.MockKGateway
import io.mockk.MockKMatcherScope
import kotlin.reflect.KClass

import kotlin.reflect.full.primaryConstructor

inline fun <reified T : Any> value(value: T): T =
    if (T::class.isInline) inlineValue(value)
    else value

@Suppress("UNCHECKED_CAST")
fun <T : Any> inlineValue(value: T): T =
    value::class.java.declaredMethods
        .find { it.name == "unbox-impl" }!!
        .invoke(value) as T

inline fun <reified T : Any> MockKMatcherScope.anyValue(): T =
    if (T::class.isInline) anyInlineValue()
    else any()

inline fun <reified T : Any> MockKMatcherScope.anyInlineValue(): T =
    T::class.primaryConstructor!!.run {
        val valueType = parameters[0].type.classifier as KClass<*>
        val any = match(ConstantMatcher<Any>(true), valueType)
        call(any)
    }

fun <T : Any> MockKMatcherScope.match(matcher: Matcher<T>, type: KClass<T>): T =
    (getProperty("callRecorder") as MockKGateway.CallRecorder).matcher(matcher, type)

val KClass<*>.isInline: Boolean
    get() = !isData &&
        primaryConstructor?.parameters?.size == 1 &&
        java.declaredMethods.any { it.name == "box-impl" }
```

**Comment by michaelbull** [comment]
@oleksiyp any chance we can get @qoomon's extension functions above shipped with mockk? Especially the `inlineValue` function is necessary any time you want to mock a field on a class that is modelled as an inline class.

**Comment by peter-kehl** [comment]
Related: Mocking calls on `inline` class instances fails, unless you use a relaxed mock:

```
inline class InlinedInt(val value: Int)

internal class TheTest {
    @Test
    fun inlinedFailsIfNotRelaxed() {
        val mockedInlinedInt = mockk<InlinedInt>()
        // --> no answer found for: InlinedInt(#1).unbox-impl()
    }

    @Test
    fun inlinedSucceedsIfRelaxed() {
        val mockedInlinedInt = mockk<InlinedInt>(relaxed = true)
    }
}
```

**Comment by peter-kehl** [comment]
We can't use `every/coEvery` with `mach` to match parameters whose value is an inlined class. Instead, use `slot and capture`:

```
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.slot
import org.junit.jupiter.api.Test
import org.assertj.core.api.Assertions.assertThat

inline class EatableUInt(val x: Int)

interface UIntEater {
    fun eat(input: EatableUInt)
}

class InlineClassParameterMockkTest {
    val eater: UIntEater = mockk()

    @Test
    fun test() {
        val wrappedValueSlot = slot<Int>()
        every {
            // As of Mockk 1.10.0 we can't use match {...} for parameters
            // whose types are inline classes. Hence capture/slot.
            eater.eat(EatableUInt(capture(wrappedValueSlot)))
        } just Runs

        eater.eat(EatableUInt(0))
        assertThat(wrappedValueSlot.captured).isEqualTo(0)
    }
}
```

**Comment by creati8e** [comment]
@peter-kehl thanks, it works!

**Comment by michaelbannister** [comment]
FWIW, I've just left a comment on the Jetbrains Kotlin issue [KT-42434](https://youtrack.jetbrains.com/issue/KT-42434) about inline classes going GA, maybe they have the resources & expertise to help figure out a tidy solution…

**Comment by spanierm** [comment]
@qoomon Thanks for your solution. It works in almost any case for me. However, I am struggling with a `null` return value for a nullable inline class. Any ideas how to deal with that one?

```text
MyId cannot be cast to class java.util.UUID (MyId is in unnamed module of loader 'app'; java.util.UUID is in module java.base of loader 'bootstrap')
java.lang.ClassCastException: MyId cannot be cast to class java.util.UUID (MyId is in unnamed module of loader 'app'; java.util.UUID is in module java.base of loader 'bootstrap')
```

**Comment by qoomon** [comment]
@spanierm may you can provide the stack trace and a test method example.

**Comment by spanierm** [comment]
@qoomon Thanks a lot for insisting on a minimal example. I built it and it worked exactly as expected. After investigating further it turned out that I missed to wrap one other inline class in another mockk stub call. So everything works as expected :)

**Comment by svenjacobs** [comment]
This has become more important now since inline classes are stable since [Kotlin 1.5.0](https://kotlinlang.org/docs/whatsnew15.html#inline-classes). I'm currently unable to use the `any()` matcher on a parameter with an inline class type.

**Comment by qoomon** [comment]
Here is my version for kotlin 1.5.0

**Usage**
```kotlin
every { dummyFunction2(days(1)) }         returns returnValueClass(days(42))
every { dummyFunction1(anyValueClass()) } returns returnValueClass(days(42))
```

**Implementation**
```kotlin
import io.mockk.ConstantMatcher
import io.mockk.MockKGateway.CallRecorder
import io.mockk.MockKMatcherScope
import kotlin.reflect.KClass
import kotlin.reflect.KProperty1
import kotlin.reflect.full.declaredMemberProperties
import kotlin.reflect.full.primaryConstructor
import kotlin.reflect.jvm.isAccessible


fun <T : Any> returnValueClass(value: T): T {
    require(value::class.isValue)
    val constructor = value::class.primaryConstructor!!
    val constructorParameter = constructor.parameters[0]
    val memberProperty = value::class.declaredMemberProperties
        .first { it.name == constructorParameter.name }
        .apply { isAccessible = true }
        .let @Suppress("UNCHECKED_CAST") { it as KProperty1<T, T> }
    return memberProperty.get(value)
}

inline fun <reified T : Any> MockKMatcherScope.anyValueClass(): T {
    require(T::class.isValue)
    val constructor = T::class.primaryConstructor!!
    val constructorParameterType = constructor.parameters[0].type.classifier as KClass<*>
    val callRecorder = getProperty("callRecorder") as CallRecorder
    val anyMatcher = callRecorder.matcher(ConstantMatcher<T>(true), constructorParameterType)
    return constructor.call(anyMatcher)
}
```

**Comment by Raibaz** [comment]
@qoomon, thanks a lot for the workaround!

However, it's related to value classes rather than inline classes, and i don't think we want to allow mocking value objects because it's widely considered to be a bad practice.

This being said, I think we can work out something similar for inline classes, although it will require using the @JvmInline annotation which has only been available since kotlin 1.5.

**Comment by qoomon** [comment]
@Raibaz inline classes or the inline modifier is deprecated since kotlin 1.5, in favour of value classes, see "Inline classes | Kotlin" https://kotlinlang.org/docs/inline-classes.html

Regarding mocking those value classes, I don't want to mock them, my solution is just a workaround to be able to mock functions with value classes as parameters or return value.

The main problem seams to be some internal mockk magic that results in comparing value classes and it raw wrapped value at runtime.
However I was not able to figure it out when and why it's happening.

**Comment by Raibaz** [comment]
Oh, got it, that makes a lot of sense, then your implementation looks perfect to be able to return value classes from mocked functions.

How would you feel about submitting a PR with it?

About comparing value classes and their wrapped values, it looks like something similar to what has been done about it in Mockito, which is being discussed in #485.

I was already thinking about doing something in the same direction, so we could probably integrate it in your implementation.

What do you think?

**Comment by qoomon** [comment]
I think I found the right spot for the fix https://github.com/mockk/mockk/blob/02a49b2f3639e066ff96d35a160bf301fe0bbb00/agent/jvm/src/main/kotlin/io/mockk/proxy/jvm/advice/Interceptor.kt#L20

before we return the value of `handler.invocation()` we need to unbox value classes like so:

```kotlin
override fun call(): Any? {
    val callOriginalMethod = SelfCallEliminatorCallable(
        MethodCall(self, method, arguments),
        self,
        method
    )

    return  handler.invocation(self, method, callOriginalMethod, arguments)?.let {
        if (value::class.isValue) unboxValueClass(it) else it
    }
}

fun unboxValueClass(value: Any): Any? {
    require(value::class.isValue)
    val constructor = value::class.primaryConstructor!!
    val constructorParameter = constructor.parameters[0]
    val memberProperty = value::class.declaredMemberProperties
        .first { it.name == constructorParameter.name }
        .apply { isAccessible = true }
        .let @Suppress("UNCHECKED_CAST") { it as KProperty1<Any, Any?> }
    return memberProperty.get(value)
}
```

**Comment by qoomon** [comment]
@Raibaz I'd like to do a PR however I'm struggling updating updating the project to kotlin 1.5. I'm not an android developer so it would be nice if you could create a branch with kotlin version 1.5, then I will create a PR with my changes and some tests.
And you probably need to update gradle version as well.

**Comment by Raibaz** [comment]
Thanks for this!

I just pushed to master some changes to make mockk compatible with 1.5.0, you should be able to make it work now.

By default it still builds with 1.3.72, to build with kotlin 1.5.0 you need to pass the version number to gradle like this:

```
./gradlew build -Pkotlin.version=1.5.0
```

I'm definitely in favor of upgrading to 1.5.0 altogether, we just need to make sure that we don't break backwards compatibility, i.e. that mockk keeps working in projects that still use 1.4.* and 1.3.*.

I hade a good number of headaches when i upgraded to 1.4 for the current version, and then somehow managed to do it with `kotlinOptions.apiVersion` and `kotlinOptions.languageVersion` that seem to be deprecated now.

It should be ok for now though, let's just make sure that the tests in `client-tests` keep passing and, if possible, extend the test coverage in that module.

**Comment by Raibaz** [comment]
Yes that's exactly what I had in mind, to use reflection to see if the 1.5 features are available in order to avoid breaking backwards compatibility.

Thanks a lot!

**Comment by qoomon** [comment]
@Raibaz I just open a PR #633 so we can work together.

**what is working**
* mockk functions with value class object as return value
* mockk functions with value class object as function argument and return value
* mockk functions with value class object as function argument and answer value

**what is NOT working**
* mockk functions with `any()` value class object as function argument

**Comment by qoomon** [comment]
@Raibaz all test are passing now :)

**Comment by qoomon** [comment]
Should work now.

**Comment by TheBestPessimist** [comment]
on mockK 1.12 this still doesn't work:


Please see the following example:

```kotlin

import io.mockk.every
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
internal class a {
    @Test
    fun `DeliveryStateMessage to Ebel`() {
        val aService = mockk<AService>()

        every { aService.getById(any()) } returns AServiceIdReturn("tralala") // this fails with NPE

        assertThat(aService.getById(AServiceId("1"))).usingRecursiveComparison()
            .isNotEqualTo(AServiceIdReturn("123"))
            .isEqualTo(AServiceIdReturn("tralala"))
    }
}

class AService {
    fun getById(id: AServiceId): AServiceIdReturn {
        return AServiceIdReturn("123")
    }
}

@JvmInline
value class AServiceId(val s: String)

@JvmInline
value class AServiceIdReturn(val s: String)
```


And the stack trace is:
```
14:04:03.871 [Test worker] DEBUG io.mockk.impl.instantiation.AbstractMockFactory - Creating mockk for AService name=#1

java.lang.NullPointerException
	at io.mockk.impl.recording.states.RecordingState.matcher(RecordingState.kt:53)
	at io.mockk.impl.recording.CommonCallRecorder.matcher(CommonCallRecorder.kt:52)
	at com.nagarro.wwsc.retailplatform.to.store.converters.a$DeliveryStateMessage to Ebel$1.invoke-Cv2eMwI(a.kt:37)
	at com.nagarro.wwsc.retailplatform.to.store.converters.a$DeliveryStateMessage to Ebel$1.invoke(a.kt:16)
	at io.mockk.impl.eval.RecordedBlockEvaluator$record$block$1.invoke(RecordedBlockEvaluator.kt:25)
	at io.mockk.impl.eval.RecordedBlockEvaluator$enhanceWithRethrow$1.invoke(RecordedBlockEvaluator.kt:78)
	at io.mockk.impl.recording.JvmAutoHinter.autoHint(JvmAutoHinter.kt:23)
	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:40)
	at io.mockk.impl.eval.EveryBlockEvaluator.every(EveryBlockEvaluator.kt:30)
	at io.mockk.MockKDsl.internalEvery(API.kt:92)
	at io.mockk.MockKKt.every(MockK.kt:98)
```

**Comment by qoomon** [comment]
@TheBestPessimist  Which kotlin version did you use?

**Comment by qoomon** [comment]
there is a test in place for `any()` that works, however it is build with kotlin <1.5 and therefore with inline class instead of value class see https://github.com/mockk/mockk/blob/master/mockk/common/src/test/kotlin/io/mockk/it/ValueClassTest.kt.

the problem ist that the initialization for an value class creates an value class object with a null value field. `io.mockk.proxy.jvm.ObjenesisInstantiator:42`

**Comment by boiler23** [comment]
In my case, this works on 1.10.6, and doesn't work with 1.12.0 (kotlin 1.5.21):
```
@JvmInline
value class SomeData(val d: Long)

interface Worker {
   suspend fun doWork(someData: SomeData): SomeData
}

class UseCase {
    var data = SomeData(0L)
    val worker = ...
    suspend fun work() {
        data = worker.doWork(data)
    }
}

// in test:
val someData = SomeData(0L)
val useCase = UseCase(...)
coEvery { worker.doWork(someData) } returns someData
useCase.work()
```

I get ClassCastException between SomeData and Long.

**Comment by dybarsky** [comment]
Hi! I have same `ClassCastException`. I am using mockk ``.12.0` and kotlin `1.5.31`.

**Comment by dybarsky** [comment]
Any updates regarding this @Raibaz ?

**Comment by aSemy** [comment]
I'm on Kotlin 1.7.0 and value classes aren't supported.

I updated the workaround [from this comment](https://github.com/mockk/mockk/issues/152#issuecomment-631796323) to use the [`isValue` flag](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect/-k-class/is-value.html), and to include the workaround for private primary constructors [shared here](https://github.com/mockk/mockk/issues/152#issuecomment-629024916).

```kotlin
import kotlin.reflect.KClass
import kotlin.reflect.full.primaryConstructor
import kotlin.reflect.jvm.isAccessible


// Work-around for mocking inline-classes
// https://github.com/mockk/mockk/issues/152#issuecomment-631796323


inline fun <reified T : Any> MockKMatcherScope.anyValue(): T =
    if (T::class.isValue) {
        anyInline()
    } else {
        any()
    }


inline fun <reified T : Any> MockKMatcherScope.anyInline(): T =
    T::class.primaryConstructor!!.run {
        this.isAccessible = true
        val valueType = parameters[0].type.classifier as KClass<*>
        call(match(ConstantMatcher(true), valueType))
    }


fun <T : Any> MockKMatcherScope.match(matcher: Matcher<T>, type: KClass<T>): T =
    (getProperty("callRecorder") as MockKGateway.CallRecorder).matcher(matcher, type)
```

I would really appreciate it if there was a similar workaround for slots and capture! I tried creating one based on [this comment](https://github.com/mockk/mockk/issues/152#issuecomment-719031430) - but it's very confusing. 

```kotlin
// doesn't work :(

inline fun <reified T : Any> MockKMatcherScope.captureValue(slot: CapturingSlot<T>): T =
    if (T::class.isValue) {
        captureInline(slot)
    } else {
        capture(slot)
    }

inline fun <reified T : Any> MockKMatcherScope.captureInline(slot: CapturingSlot<T>): T =
    T::class.primaryConstructor!!.run {
        this.isAccessible = true
        val valueType = parameters[0].type.classifier as KClass<*>
        val valueMatcher = CapturingSlotMatcher(CapturingSlot(), valueType)
        val valueSlot = valueMatcher.captureSlot

        call(match(CapturingSlotMatcher(slot, T::class), valueType)
    }
```

**Comment by qoomon** [comment]
@aSemy I managed to implement a working `captureValue` function. See full example below.
```kotlin
inline fun <reified T : Any> MockKMatcherScope.anyValue(): T {
    if (!T::class.isValue) return any()

    val constructor = T::class.primaryConstructor!!.apply { isAccessible = true }
    val rawType = constructor.parameters[0].type.classifier as KClass<*>

    val anyRawValue = callRecorder.matcher(ConstantMatcher<T>(true), rawType)
    return constructor.call(anyRawValue)
}

inline fun <reified T : Any> MockKMatcherScope.captureValue(slot: CapturingSlot<T>): T {
    if (!T::class.isValue) return capture(slot)

    val constructor = T::class.primaryConstructor!!.apply { isAccessible = true }
    val rawType = constructor.parameters[0].type.classifier as KClass<*>

    val anyRawValue = callRecorder.matcher(CapturingValueSlotMatcher(slot, constructor, rawType), rawType)
    return constructor.call(anyRawValue)
}

val MockKMatcherScope.callRecorder: MockKGateway.CallRecorder
    get() = getProperty("callRecorder") as MockKGateway.CallRecorder

data class CapturingValueSlotMatcher<T : Any>(
    val captureSlot: CapturingSlot<T>,
    val valueConstructor: KFunction<T>,
    override val argumentType: KClass<*>,
) : Matcher<T>, CapturingMatcher, TypedMatcher, EquivalentMatcher {
    override fun equivalent(): Matcher<Any> = ConstantMatcher(true)

    @Suppress("UNCHECKED_CAST")
    override fun capture(arg: Any?) {
        if (arg == null) {
            captureSlot.isNull = true
        } else {
            captureSlot.isNull = false
            captureSlot.captured = valueConstructor.call(arg)
        }
        captureSlot.isCaptured = true
    }

    override fun match(arg: T?): Boolean = true

    override fun toString(): String = "slotCapture<${argumentType.simpleName}>()"
}
```

**Comment by aSemy** [comment]
Great! That looks really tidy. Can we get this committed to the main repo? If you'd like I can take your code and start a PR with it @qoomon?

**Comment by qoomon** [comment]
I'm about to implement it and create a PR.

**Comment by qoomon** [comment]
@aSemy I think I need help. I can't get the project working. So feel free to create a PR. BTW here are some tests already.
```kotlin
package me.qoomon.enhancements.mockk

import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import org.junit.jupiter.api.Test
import strikt.api.expectThat
import strikt.assertions.isEqualTo

class MockkTest {

    @Test
    fun `any matcher for value class`() {
        // given
        val mock = mockk<ValueServiceDummy>(relaxed = true)
        val givenResult = 1
        every { mock.doSomething(anyValue()) } returns givenResult

        // when
        val result = mock.doSomething(ValueDummy("moin"))

        // then
        expectThat(result).isEqualTo(givenResult)
    }

    @Test
    fun `slot for value class`() {
        // given
        val mock = mockk<ValueServiceDummy>(relaxed = true)
        val slot = slot<ValueDummy>()
        val givenResult = 1
        every { mock.doSomething(captureValue(slot)) } returns givenResult

        val givenParameter = ValueDummy("s")

        // when
        val result = mock.doSomething(givenParameter)

        // then
        expectThat(result).isEqualTo(givenResult)
        expectThat(slot.captured).isEqualTo(givenParameter)
    }

    @Test
    fun `value class as return value`() {
        // given
        val mock = mockk<ValueServiceDummy>(relaxed = true)
        val givenResult = ValueDummy("moin")
        every { mock.getSomething() } returns givenResult

        // when
        val result = mock.getSomething()

        // then
        expectThat(result).isEqualTo(givenResult)
    }
}

@JvmInline
value class ValueDummy(val value: String)

interface ValueServiceDummy {
    fun doSomething(value: ValueDummy): Int
    fun getSomething(): ValueDummy
}
```

**Comment by qoomon** [comment]
@aSemy 
```kotlin
/**
 * POLYFILL for kotlin version < 1.5
 * will be shadowed by implementation in kotlin SDK 1.5+
 *
 * @return true if this is an inline class, else false
 */
private val <T : Any> KClass<T>.isValue: Boolean
    get() = !isData &&
            primaryConstructor?.parameters?.size == 1 &&
            java.declaredMethods.any { it.name == "box-impl" }
```

**Comment by dybarsky** [comment]
Thank you <3

**Comment by equeim** [comment]
Doesn't seem to work in my case with coEvery :(

```
Failed matching mocking signature for
SignedCall(retValue=, isRetValueMock=true, retType=class kotlin.collections.List, self=DonkiRepositoryInternal(#52), method=updateEventsForWeek-_A-dL5c(LocalDate, EventType, Continuation), args=[0000-00-00, null, Continuation at org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest$Some of initial load weeks require refresh$3.invokeSuspend(EventsSummariesRemoteMediatorTest.kt:89)], invocationStr=DonkiRepositoryInternal(#52).updateEventsForWeek-_A-dL5c(0000-00-00, null, continuation {}))
left matchers: [any()]
io.mockk.MockKException: Failed matching mocking signature for
SignedCall(retValue=, isRetValueMock=true, retType=class kotlin.collections.List, self=DonkiRepositoryInternal(#52), method=updateEventsForWeek-_A-dL5c(LocalDate, EventType, Continuation), args=[0000-00-00, null, Continuation at org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest$Some of initial load weeks require refresh$3.invokeSuspend(EventsSummariesRemoteMediatorTest.kt:89)], invocationStr=DonkiRepositoryInternal(#52).updateEventsForWeek-_A-dL5c(0000-00-00, null, continuation {}))
left matchers: [any()]
	at app//io.mockk.impl.recording.SignatureMatcherDetector.detect(SignatureMatcherDetector.kt:99)
	at app//io.mockk.impl.recording.states.RecordingState.signMatchers(RecordingState.kt:39)
	at app//io.mockk.impl.recording.states.RecordingState.round(RecordingState.kt:31)
	at app//io.mockk.impl.recording.CommonCallRecorder.round(CommonCallRecorder.kt:50)
	at app//io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:63)
	at app//io.mockk.impl.eval.EveryBlockEvaluator.every(EveryBlockEvaluator.kt:30)
	at app//io.mockk.MockKDsl.internalCoEvery(API.kt:99)
	at app//io.mockk.MockKKt.coEvery(MockK.kt:116)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest.Some of initial load weeks require refresh(EventsSummariesRemoteMediatorTest.kt:89)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest.access$Some of initial load weeks require refresh(EventsSummariesRemoteMediatorTest.kt:26)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest$Some of initial load weeks require refresh ## validate initialize()$1.invokeSuspend(EventsSummariesRemoteMediatorTest.kt:95)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest$Some of initial load weeks require refresh ## validate initialize()$1.invoke(EventsSummariesRemoteMediatorTest.kt)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest$Some of initial load weeks require refresh ## validate initialize()$1.invoke(EventsSummariesRemoteMediatorTest.kt)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest$base$1.invokeSuspend(EventsSummariesRemoteMediatorTest.kt:247)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest$base$1.invoke(EventsSummariesRemoteMediatorTest.kt)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest$base$1.invoke(EventsSummariesRemoteMediatorTest.kt)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTestCoroutine$2.invokeSuspend(TestBuilders.kt:212)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTestCoroutine$2.invoke(TestBuilders.kt)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTestCoroutine$2.invoke(TestBuilders.kt)
	at app//kotlinx.coroutines.intrinsics.UndispatchedKt.startCoroutineUndispatched(Undispatched.kt:55)
	at app//kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:112)
	at app//kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTestCoroutine(TestBuilders.kt:211)
	at app//kotlinx.coroutines.test.TestBuildersKt.runTestCoroutine(Unknown Source)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$1$1.invokeSuspend(TestBuilders.kt:167)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$1$1.invoke(TestBuilders.kt)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$1$1.invoke(TestBuilders.kt)
	at app//kotlinx.coroutines.test.TestBuildersJvmKt$createTestResult$1.invokeSuspend(TestBuildersJvm.kt:13)
	at app//kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at app//kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at app//kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at app//kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at app//kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at app//kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at app//kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at app//kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at app//kotlinx.coroutines.test.TestBuildersJvmKt.createTestResult(TestBuildersJvm.kt:12)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest(TestBuilders.kt:166)
	at app//kotlinx.coroutines.test.TestBuildersKt.runTest(Unknown Source)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest(TestBuilders.kt:154)
	at app//kotlinx.coroutines.test.TestBuildersKt.runTest(Unknown Source)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest$default(TestBuilders.kt:147)
	at app//kotlinx.coroutines.test.TestBuildersKt.runTest$default(Unknown Source)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest.base(EventsSummariesRemoteMediatorTest.kt:244)
	at app//org.equeim.spacer.donki.data.paging.EventsSummariesRemoteMediatorTest.Some of initial load weeks require refresh ## validate initialize()(EventsSummariesRemoteMediatorTest.kt:94)
	at java.base@11.0.13/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.13/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.13/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.13/java.lang.reflect.Method.invoke(Method.java:566)
	at app//org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at app//org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at app//org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at app//org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at app//org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at app//org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at app//org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at app//org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at app//org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at app//org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at app//org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at app//org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at app//org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at app//org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at app//org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at app//org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at app//org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at app//org.junit.runners.Suite.runChild(Suite.java:128)
	at app//org.junit.runners.Suite.runChild(Suite.java:27)
	at app//org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at app//org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at app//org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at app//org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at app//org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at app//org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at app//org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base@11.0.13/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.13/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.13/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.13/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

Week is declared as
```kotlin
@JvmInline
internal value class Week @VisibleForTesting constructor(
    val firstDay: LocalDate
) : Comparable<Week>
```

It is passed to suspend function directly as simple parameter.

Kotlin version is 1.7.10

**Comment by qoomon** [comment]
I've created further failing tests.

```kotlin
class MockkTest {

    interface ServiceDummy {
        fun getAny(): Any
        fun getResult(): Result<Int>
        fun execute(block: ServiceDummy.() -> Int)
    }

    @Test
    fun `return Result for function with Result return type`() {
        val givenValue = Result.success(42)
        val service = mockk<ServiceDummy> {
            every { getResult() } returns givenValue
        }

        val result = service.getResult()

        assertEquals(givenValue, result)
        verify { dummy.getResult() }
    }

    @Test
    fun `return Result for function with Any return type`() {
        val givenValue = Result.success(42)
        val service = mockk<ServiceDummy> {
            every { getAny() } returns givenValue
        }

        val result = service.getAny()

        assertEquals(givenValue, result)
        verify { dummy.getAny() }
    }

    @Test
    fun `pass extension blocks as parameter`() {
        val givenValue = 42
        val dummy = mockk<ServiceDummy> {
            every { execute(any()) } answers {
                val block = arg<ServiceDummy.() -> Int>(0)
                this@mockk.block()
            }
        }

        val result = dummy.execute { givenValue }
        
        assertEquals(givenValue, result)
        verify { dummy.execute(any()) }
    }
}
```

**Comment by qoomon** [comment]
@aSemy  I've created further tests, however I was not able fix it in away so that all test are passing. There are quite some location where value classes need to be handled or already be handled. Maybe you are capable to fix these tests.

```kotlin
class MockkTest {

    interface ServiceDummy {
        fun getAny(): Any
        fun getResult(): Result<Int>
        fun execute(block: ServiceDummy.() -> Int)
    }

    @Test
    fun `return Result for function with Result return type`() {
        val givenValue = Result.success(42)
        val service = mockk<ServiceDummy> {
            every { getResult() } returns givenValue
        }

        val result = service.getResult()

        assertEquals(givenValue, result)
        verify { dummy.getResult() }
    }

    @Test
    fun `return Result for function with Any return type`() {
        val givenValue = Result.success(42)
        val service = mockk<ServiceDummy> {
            every { getAny() } returns givenValue
        }

        val result = service.getAny()

        assertEquals(givenValue, result)
        verify { dummy.getAny() }
    }

    @Test
    fun `pass extension blocks as parameter`() {
        val givenValue = 42
        val dummy = mockk<ServiceDummy> {
            every { execute(any()) } answers {
                val block = arg<ServiceDummy.() -> Int>(0)
                this@mockk.block()
            }
        }

        val result = dummy.execute { givenValue }
        
        assertEquals(givenValue, result)
        verify { dummy.execute(any()) }
    }
}
```

**Comment by aSemy** [comment]
thanks @qoomon. I'll look into this after #855 is done. I suspect something like the fix I proposed here https://github.com/mockk/mockk/pull/849#issuecomment-1193278216 is needed, to ensure consistent behaviour.

I've also been pondering using some code generation (like Kotlin Poet) to generate tests for ALL possible combinations of value class args/responses, nullables, mock types, matcher types...

**Comment by qoomon** [comment]
> I've also been pondering using some code generation (like Kotlin Poet) to generate tests for ALL possible combinations of value class args/responses, nullables, mock types, matcher types...

That sounds interesting. I would be excited to see your solution.

**Comment by Raibaz** [comment]
Yep that'd be pretty cool!

**Comment by fgubler** [comment]
Just to mention it: the workarounds mentioned above don't work if the Value-Class has a generic parameter.
For example
```
@JvmInline
value class NonEmptyList<out T> private constructor(val value: List<T>) : List<T> by value

fun foo(list: NonEmptyList<String>)
```
=> The method `every { foo(anyValue()) } just runs` will fail with the exception 
"io.mockk.MockKException: Failed matching mocking signature for"

**Comment by j1c1m1b1** [comment]
Unfortunately, there is still an issue with this feature. If you have an instance that extends from a class with generic values, if the output value of an execution of a function is a value class instance a `ClassCastException` is thrown:

In this case `ProductToProductInformationUiModelMapper` extends from:

```kotlin
abstract class Mapper<in T, out R> {
    suspend fun map(parameter: T) : R
}
```

In this case `R` will be:

```kotlin
value class ProductItemUiModel(value: String) {
// Some value class validation
}
```

```kotlin
private val productToProductInformationUiModelMapper: ProductToProductInformationUiModelMapper = mockk()

private val thingToTest: ThingToTest = ThingToTest(productToProductInformationUiModelMapper)

@Test
fun `test somthing`() {
    val expectedValue = getSomeExpectedValue()
    val productItemUiModelValue = expectedValue.productItemUiModelValue
    val product = generateRandomProduct()

    coEvety { productToProductInformationUiModelMapper.map(product) } returns 
    ProductItemUiModel(productItemUiModelValue)

    runTest {
         val actualValue = thintToTest.doSomethingCool(product)
    }
}
```

Running this test throws:

```
class java.lang.String cannot be cast to class com.some.package.ProductInformationUiModel
```

**Comment by neiser** [comment]
@j1c1m1b1 Maybe it's worthwhile to open a separate issue, as your comment might have been missed by the maintainers?
